### ISSUE TITLE
How to dump model formula ?

### ISSUE THREAD

**Issue by fcoiffie** [jcrudy/py-earth#63]
Is there a way to dump the model formula ?

**Comment by ColCarroll** [comment]
I have done a little work to provide access to model coefficients, but got bogged down in trying to decide a good data structure to put them in.  `model.coef_` should return an array, but in this case it is dependent on where in the feature space you are asking.Perhaps `model.coef_(x)`?  Or maybe `model.coef_` returns a dictionary or arrays?  Or there will be a separate array describing the boundaries of each row from `model.coef_`?

**Comment by ColCarroll** [comment]
I should add that the relevant commit is here:
https://github.com/ColCarroll/py-earth/commit/9640b593008d7db48c00aedee3552de3e2a679fd

and it gives a method `labeled_coef_`, whose output looks something like this (this is time series data, using the `xlabels` argument when fitting the model):
```
[{'(Intercept)': 23.331373525526406}, 
{'h(Date-899)': 0.039051102948551353}, 
{'h(Date-1186)': -1.4419994444986672}, 
{'h(Date-996)': 0.79201279381621625}, 
{'h(996-Date)': -0.80098522842398767}, 
{'h(Date-1254)': 0.98487709201589901}, 
{'h(Date-1547)': -1.2262192178089002}, 
{'h(1010-Date)': 0.78886444153185165}]
```

**Comment by fcoiffie** [comment]
@ColCarroll Yes, it will help me. But in my case, instead of having usual hinge function 'h(x-constant)', I get 'C(x0|s=+1,200,350,525)'... And I don't know how to interpret this term...

**Comment by fcoiffie** [comment]
I understand : it's due to "smooth=True".

**Comment by fcoiffie** [comment]
And by formula, I expected something like model = 23.331373525526406 + 0.039051102948551353 * max(0, Date-899) -1.4419994444986672 * max(0, Date-1186) ... (My purpose is to reuse the model in a software code)

**Comment by ColCarroll** [comment]
Your best bet for that specific output might be to either use that fork, or turn the method into a function, and use string formatting.  If @jcrudy has any suggestions for what he thinks the model parameters should look like, I'd be happy to submit a pull request (as this addresses issue #4).

**Comment by jcrudy** [comment]
@ColCarroll @fcoiffie Thanks for the input and code.  Here are some ideas for output formats.  There's no reason there can't be more than one of these:

1. Valid Python code
2. A sympy expression
3. A Python function
4. A PMML object (see augustus)
5. An ElementTree representing a PMML model
6. A human readable string
7. Valid code in languages other than Python (Java, C, etc.)

I'd appreciate feedback on these or any additional suggestions.  I'd also welcome implementations, but let's talk more before if you're going to spend significant time on it.  I have to think about how exactly the interface should work.  Probably just a method for each output format.  Something like to_py_code(), to_sympy(), to_py_func(), that sort of thing.  I should look at whether scikit-learn has something similar and whether this would impact compatibility.

**Comment by fcoiffie** [comment]
@ColCarroll @jcrudy For my personal need, my request was inspired by this package : https://orange-multitarget.readthedocs.org/en/0.9.2/Orange.regression.earth.html
(I copy/paste the formula in my C code and I implement the max() with a macro but I agree this is not a generic need but it should work with many languages like Python, C, C++, Java, etc... I like earth algorithm because it fits well for embedded software as it requires only additions, multiplications and tests)

**Comment by jcrudy** [comment]
@fcoiffie That makes sense to me.  Would an option for right now be to write a new macro for the hinge function h(x)?

**Comment by fcoiffie** [comment]
@jcrudy Yes, I can define a h(x) macro. But, the hinge function definition must be known by the user (or can be added in the py-earth doc). Even the Wikipedia page uses the max() notation in these examples : https://en.wikipedia.org/wiki/Multivariate_adaptive_regression_splines#The_basics

**Comment by jcrudy** [comment]
@fcoiffie Oh, yes, I absolutely agree.  Just wanted to make sure you have an option to move forward until I get my act together.  Py-earth is currently undergoing some changes related to scikit-learn compatibility, so it might be a while before I start adding new methods for exporting models.  I will start thinking about it, though.  And pull requests are definitely welcome. @ColCarroll 

@mehdidc Do you have any comment on how model export methods discussed in this thread would impact sklearn compatibility?

**Comment by mehdidc** [comment]
@jcrudy sorry for late answer. To my knowledge, it's preferable to make all the methods private except the standard API methods. I would say one way to do this is to add functions export_* in earth.py like it is done in sklearn.tree.export_graphviz : http://scikit-learn.org/stable/modules/generated/sklearn.tree.export_graphviz.html

**Comment by jcrudy** [comment]
@mehdidc Thanks, that's a helpful example.  So, if I am understanding correctly, the sklearn compliant way to have this functionality would be to have the public interface be a separate function rather than a method of the Earth class itself.

**Comment by mangecoeur** [comment]
I'm also interested in this issue - I'm struggling to figure out exactly how to extract the knot positions an linear segment gradients from the model result.

Also, from the scikit doc there isn't really a pattern for what you describe since there is just the one function. That said it would be logical for the model conversion function to be external to the model itself, that way you can implement any number of representations (as text, as parameters lists, etc) without clogging up the Model class with every kind of converter...

**Comment by jcrudy** [comment]
Just merged pull request #72 by @ColCarroll, which addresses this issue.  There are still plenty of other formats that could be exported into, but for now I think I will consider this issue closed.  If anyone has suggestions for other formats to export to, please feel free to open a new issue or pull request.

@mangecoeur Does @ColCarroll's work cover your use case?
